runtime/domain.c

```diff
@@ -1070,6 +1076,38 @@ CAMLexport _Atomic caml_timing_hook caml_domain_terminated_hook =
 
 static void domain_terminate(void);
 
+static void domain_sync_result(value term_sync, value res_or_exn)
+{
+  CAMLparam1(term_sync);
+  CAMLlocal1(res);
+  if (Is_exception_result(res_or_exn)) {
+    res = Extract_exception(res_or_exn);
+    /* [Finished (Error res)] from stdlib/domain.ml */
+    res = caml_alloc_1(1, res);
+    res = caml_alloc_1(0, res);
+  } else {
+    res = res_or_exn;
+  }
+
+  /* Synchronize with joining domains. We call [caml_ml_mutex_lock]
+     because the systhreads are still running on this domain. We
+     assume this does not fail the exception it would raise at this
+     point would be bad for us. */
+  caml_ml_mutex_lock(*Term_mutex(term_sync));
+
+  /* Store result */
+  volatile value *state = Term_state(term_sync);
+  CAMLassert(!Is_block(*state));
```

Comment:
Why not assert directly that `*state` is `Running`?

Reply:
I simply adapted the test that was made before.